bluesky/bundlers.py

```diff
@@ -595,68 +626,293 @@ async def complete(self, msg):
+                "`StreamRange(start=0, stop=0)` on `ComposeStreamDatum` when used with the "
+                "run engine."
+            )
+
+        indices_difference = doc["indices"]["stop"] - doc["indices"]["start"]
+
+        if (
+            stream_datum_previous_indices_difference and
+            stream_datum_previous_indices_difference != indices_difference
+        ):
+            raise EventModelValueError(
+                f"Received `indices` {doc['indices']} during `collect()` these are of a different "
+                f"width `{indices_difference}` than other detectors in the same collect() or save()."
+            )
+
+        index_in_chunk = next(stream_datum_data_key_counter)
+
+        current_seq_counter = self._sequence_counters[message_stream_name]
+
+        doc["seq_nums"] = StreamRange(
+            start=current_seq_counter, stop=current_seq_counter + indices_difference
+        )
+
+        if index_in_chunk == 0:
+            self._sequence_counters[message_stream_name] += indices_difference
+
+            # Return a prevous width of 0 to allow the next stream_datum to have a new width
+            return 0
+
+        return indices_difference
+
+    async def _pack_external_assets(
+        self,
+        asset_docs: Iterable[Tuple[str, ExternalAssetDoc]],
+        message_stream_name: Optional[str]
+    ):
+        """Packs some external asset documents with relevant information from the run."""
+
+        stream_resource_data_keys = []
+        stream_datum_data_key_counter = None
+        stream_datum_previous_indices_difference = 0
+
+        for name, doc in asset_docs:
+            if name == DocumentNames.resource.value:
+                doc["run_start"] = self._run_start_uid
+            elif name == DocumentNames.stream_resource.value:
+                if stream_datum_data_key_counter:
+                    raise RuntimeError("Received a `stream_resource` document after `stream_datum` documents.")
+
+                if doc["data_key"] in stream_resource_data_keys:
+                    raise RuntimeError(f"Received a second `stream_resource` with a `data_key` {doc['data_key']}.")
+
+                doc["run_start"] = self._run_start_uid
+
+                # Information required to pack stream_datum seq_nums
+                stream_resource_data_keys.append(doc["data_key"])
+
+            elif name == DocumentNames.stream_datum.value:
+                if not stream_resource_data_keys:
+                    raise RuntimeError(
+                        "Received `stream_datum` without recieving any `stream_resource`"
+                    )
+                if not stream_datum_data_key_counter:
+                    # Create a counter of data_keys for `stream_datum` `indices` to be validated against
+                    counter_list = list(range(len(stream_resource_data_keys)))
+                    counter_list.reverse()
+                    stream_datum_data_key_counter = cycle(counter_list)
```

Comment:
I'm not sure what this counter is doing. I was intending that the only counter we use is `self._sequence_counter`. When we pack external assets then we should check that all StreamDatums have the same indices_difference, then assign `StreamRange(self._sequence_counter[stream_name], self._sequence_counter[stream_name] + indices_difference)` to each of them, finally doing `self._sequence_counter[stream_name] += indices_difference`. (It might be more complicated than that because we should support producing Events at the same time, but you should still be able to use the same sequence_counters`). 

I might have misunderstoon the purpose of this counter though...

Reply:
It is to validate that the indices passed in are of equal width for all the data_keys:

collect is called then we have four `stream_resource`
`stream_resource_1 + stream_resource_2 + stream_resource_3 + stream_resource_4`

then the asset docs iterator will spit out `stream_datum` for each stream resource:
```
   stream_datum_1_1 + stream_datum_2_1 + stream_datum_3_1 + stream_datum_4_1
 + stream_datum_1_2 + stream_datum_2_2 + stream_datum_3_2 + stream_datum_4_2
 + ... 
 + stream_datum_1_n + stream_datum_2_n + stream_datum_3_n + stream_datum_4_n
```


`stream_datum_m_l` is the `l`-th stream_datum for the `m`-th data_key. This counter checks to see how many contiguous stream datum we have left for each data_key before we receive a new stream_datum for the first data key: We want to check the indices difference is the same for `stream_datum_m_l` and `stream_datum_(m-1)_l`, but then allow the next stream datum `stream_datum_1_(l+1)` to have a different indices difference.

Reply:
Discussed in person. We can simplify because we decided `l` will be precisely one for every call to `collect_asset_docs`, so we can check the lengths are equal for all stream_datums and assign them seq_nums accordingly

Reply:
Never mind, this was for the old plan.